hyperspy/axes.py

```diff
@@ -598,12 +599,14 @@ class DataAxis(BaseDataAxis):
 
     def __init__(self,
                  index_in_array=None,
+                 _type=t.Undefined,
```

Comment:
What about adding ``**kwargs`` instead? Then the constructor could check if the `_type` corresponds with the class and raise a ``ValueError`` if it doesn't.

Reply:
As I understand, `_type` is not used but it is here to allow creating axis from the output of `get_axis_dictionary()`. However, this doesn't work very well, because we need to know the class before hand and instead it is better to use the `_create_axis` function to create an axis. Would it be better to remove the `_type` of all the constructor and make the `_create_axis` function public and use this function when creating axis from the `get_axis_dictionary()`? It will need a deprecation warning.

Using the `**kwargs` approach may be a bit annoying with `FunctionalDataAxis` since it will need to be removed before passing it the `Expression` component.

Reply:
For hyperspy itself, and axis initialization in particular, `_type` is not really needed. `_create_axis` recognizes the desired type of axis from the attributes (expression -> FDA, axis -> DA, else UDA). At the moment, the public version is `create_axes`, which takes a list of dictionaries instead of a single one. So the introduction of `_type` is only to ensure that a saved `.hspy` file would contain that bit of information (and that it can be more easily accessed than via `__class__.__name__`).

Having it in the `**kwargs` allows for a consistency check between the automatically defined type and the given one.  However, when e.g. converting between axes types, it then has to be removed from the dictionary or explicitly set to the correct type. As it is right now, any given `_type` would simply be ignored and the automatically determined one is instead used when a new axis is created.

Reply:
We agree on the approach but I am not convinced that we should have the `_type` attribute since it is not used; it is necessary only when exporting an axis as a dictionary, so it should be defined only then.

Also `create_axes` is a method of the `AxesManager` class and I think it may be still convenient to create an axis independently of the `AxesManager` using a public method. I don't have any particular use case in mind, but creating axis from `DataAxis` is used in two places in hyperspy.

Reply:
Good point, otherwise it is just an alias for `__class__.__name__`.